src/humanloop/otel/exporter.py

```diff
@@ -0,0 +1,305 @@
+                    self._upload_queue.put(
+                        (
+                            span,
+                            copy.deepcopy(evaluation_context),
+                        )
+                    )
+            # Reset the EvaluationContext so run eval does not
+            # create a duplicate Log
+            if evaluation_context is not None and is_evaluated_file(
+                spans[0],
+                evaluation_context,
+            ):
+                # Mark the EvaluationContext as used
+                self._client.evaluation_context_variable.set(None)
+            return SpanExportResult.SUCCESS
+        else:
+            logger.warning("HumanloopSpanExporter is shutting down, not accepting new spans")
+            return SpanExportResult.FAILURE
+
+    def shutdown(self) -> None:
+        self._shutdown = True
+        for thread in self._threads:
+            thread.join()
+
+    def force_flush(self, timeout_millis: int = 3000) -> bool:
+        self._shutdown = True
+        for thread in self._threads:
+            thread.join(timeout=timeout_millis)
+        self._upload_queue.join()
+
+        return True
+
+    def _do_work(self):
+        """Upload spans to Humanloop.
+
+        Ran by worker threads. The threads use the self._shutdown flag to wait
+        for Spans to arrive. Setting a timeout on self._upload_queue.get() risks
+        shutting down the thread early as no Spans are produced e.g. while waiting
+        for user input into the instrumented feature or application.
+
+        Each thread will upload a Span to Humanloop, provided the Span has all its
+        dependencies uploaded. The dependency happens in a Flow Trace context, where
+        the Trace parent must be uploaded first. The Span Processor will send in Spans
+        bottoms-up, while the upload of a Trace happens top-down. If a Span did not
+        have its span uploaded yet, it will be re-queued to be uploaded later.
+        """
+
+        # Do work while the Exporter was not instructed to
+        # wind down or the queue is not empty
+        while self._upload_queue.qsize() > 0 or not self._shutdown:
+            try:
+                thread_args: tuple[ReadableSpan, EvaluationContext]  # type: ignore
+                # Don't block or the thread will never be notified of the shutdown
+                thread_args = self._upload_queue.get(
+                    block=False,
+                )  # type: ignore
+                span_to_export, evaluation_context = thread_args
+                # Set the EvaluationContext for the thread so the .log action works as expected
+                # NOTE: Expecting the evaluation thread to send a single span so we are
+                #   not resetting the EvaluationContext in the scope of the export thread
+                self._client.evaluation_context_variable.set(
+                    copy.deepcopy(evaluation_context),
```

Comment:
why copying here again

Reply:
ack changed

Reply:
@harry-humanloop can you resolve please?